Wemo binding : Added new devices - Wemo Bridge and

hmerk opened the pull request:
Change autodiscovery to UPnP
Introduce GENA event usage
See discussion at #662 and #461

This addition also needs jupnp PR41 to be merged.

Signed-off-by: Hans-Jörg Merk hans-joerg.merk@t-online.de

sjka commented on bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/UpnpDiscoveryService.java:
```diff
@@ -102,6 +103,7 @@ protected void startScan() {
         }
         upnpService.getRegistry().addListener(this);
         upnpService.getControlPoint().search();
+        upnpService.getControlPoint().search(new RootDeviceHeader());
```
As I'm not an UPnP expert: Are these two calls by intention?

hmerk replied:
Yes, the second call needs to be done to find the WeMos, as they don't respond to SSDP:all search.

kaikreuzer replied:
I think we discussed that a while ago already: I would not want this (WeMo specific) call to be in the general UPnP discovery - I am pretty sure that it increases the network traffic quite a bit, which is not needed for any user that does not use WeMo devices.
I'd therefore suggest to move this call to a Wemo DiscoveryService, which does nothing else than triggering this (and the UPnPDiscoveryParticipant will then receive the results).

hmerk replied:
Ok, will be changed

sjka commented on extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml:
```diff
@@ -126,5 +154,25 @@
+        <category>DimmableLight</category>
+    </channel-type>
+
+    <!-- Color Channel -->
```
I can't see where this channel is used, so just to make sure: Is a reference to it missing anywhere?

hmerk replied:
You are wroght, this is a leftover, will delete it.

sjka commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java:
```diff
@@ -0,0 +1,274 @@
+                    + "</DevUDN><ReqListType>PAIRED_LIST</ReqListType>" + "</u:GetEndDevices>" + "</s:Body>"
+                    + "</s:Envelope>";
+
+            URL descriptorURL = service.getDescriptorURL(this);
```
You are accepting NULL for the upnpIOService, just logging it on debug level. In that case, I see a potential NPE here

sjka replied:
Oh, I just noticed that the only caller actually declared a 1..1 static dependency to it, so that this case indeed cannot happen currently

sjka commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java:
```diff
@@ -0,0 +1,274 @@
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stop WeMo device background discovery");
+
```
Wouldn't it be the idea to stop the scanningJob here?

hmerk replied:
Right, i will check.

sjka commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java:
```diff
@@ -56,151 +52,140 @@
+                    String wemoURL = getWemoURL("basicevent");
+
+                    if (wemoURL != null) {
+                        String wemoCallResponse = WemoHttpCall.executeCall(wemoURL, soapHeader, content);
```
The assigned variable is never used

hmerk replied:
Right, will change it.

sjka commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java:
```diff
@@ -0,0 +1,274 @@
+
+        private WemoLinkDiscoveryService service;
+
+        public WemoLinkScan(WemoLinkDiscoveryService service) {
```
The WemoLinkDiscoveryService variable & field are never used

hmerk replied:
Hm, i will check this.

sjka commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoLightHandler.java:
```diff
@@ -0,0 +1,363 @@
+                        break;
+                }
+                break;
+        }
```
I see that the light currently only has these two channels. However, I think it's good practice to handle the the "default" case gracefully too, i.e. either log an error and return or throw an IllegalArgumentException. In that way it will fail fast in case somebody adds another channel later on...

hmerk replied:
As you can seee, the binding only supports the original WeMo LED lights which have just these two channels. I do not recommend to support other lights cause of general reliability of WeMo devices.

sjka commented:
Thanks for the cleanup of the PR. 
One general question, as I don't have any wemo devices here: Did you have the chance yet to try it out with the hardware?

hmerk commented:
@sjka, thanks for the review. I added some comments as well and incorporate the needed changes.
Could be that the lightDiscovery will change a bit more.
And there will be added support for the WeMo maker.
To answer your question, only the lightswitch was not tested b myself, cause it's not available in Germany. But i am using the same code as in 1.x binding, where it is reported working.

I ill update this pr until next week.

hmerk commented:
@sjka I have incorporated your comments and also added the new channels for InsightSwitches and support for WemoMaker.

Best and Merry Christmas
Hans-Jörg

kaikreuzer commented:
@hmerk With https://github.com/eclipse/smarthome/pull/946, the target platform contains the new JUPnP release, which includes jupnp PR41.
Could you please test your PR against this and rebase it, if all is fine?

maggu2810 commented:
If we already know that a new JUPnP version is necessary, shouldn't we change the manifest and add a version range. So the OSGi framework knows that all JUPnP versions up to x does not fulfil the requirements of that bundle.

hmerk commented:
@kaikreuzer  just tested with the new target platform and my additions work as expected.
Also sqashed my commits if that was what you asked me for.

maggu2810 commented:
@hmerk There are conflicts. Have you rebased your commit, too?
@kaikreuzer What do you think about the version constraints for jUPnP in the Manifest?

hmerk commented:
Yes I did, but maybe I made a mistake.
As I still got a version before I squashed the commits, can you give me a hint what the right procedure would be? So I could try again.

maggu2810 commented:
Sometimes people get in trouble with all that rebase, squash etc. stuff. Perhaps you should save your a copy of your branch somewhere (e.g. to another branch).

I would do something similar to:
```bash
git fetch <remote-eclipse>
git rebase <remote-eclipse>/master
```

kaikreuzer commented:
> @kaikreuzer What do you think about the version constraints for jUPnP in the Manifest?

I am rather indifferent here. We do not maintain version constraints for any other 3rd party libs, so I am not sure if we should start with it right here. The normal assumption is that the TP is the reference for required versions.

maggu2810 commented:
For persons that read all the issues and pull requests it is rather easy to adapt their target platform (of the product / solution) after such a code change. Perhaps we could expect that everyone that is using Eclipse SmartHome knows about such changes.
I am also fine not using version constraints just because the Eclipse IDE is using the current version of the target platform as a minimum versions.
But if we already know that there is a minimum version and using a lower version will result in a not working bundle, I would like to add such an information.
Isn't that the intention of the OSGi package imports and their version ranges (or of the whole OSGi stuff), that the OSGi runtime uses this information to decide if a bundle could be resolved? All that manifest approach is used to find missing constraints at resolution and not at runtime.

Sure, we cannot force that everyone adds this information and start to test what is the minimum version of a dependency.
I do not like to change anything about the current way we handle this! But if we know such a minimum version already, I would prefer to add it to the manifest and suffer good OSGi bundles / bundle manifests.

maggu2810 commented:
In this case here I do not know if a version 'RC1' is recognized correctly by the OSGi version comparison.

kaikreuzer commented:
> But if we already know that there is a minimum version and using a lower version will result in a not working bundle, I would like to add such an information.

I am ok with that.

> In this case here I do not know if a version 'RC1' is recognized correctly by the OSGi version comparison.

Since RC1 is the only version available in the TP, there should be no issues.

hmerk commented:
Ok, finally got it rebased and squashed.

kaikreuzer commented:
Congrats! So how about adding the version constraint for JUPnP as @maggu2810 suggested in https://github.com/eclipse/smarthome/pull/721#issuecomment-177640911?

hmerk commented:
@kaikreuzer 
@maggu2810 
I don't really like the idea of using version constraints. Just in this case we had a change of the jupnp version, but what if there are changes needed before a version will change? How to make sure, one uses the right "version" in that case?

kaikreuzer commented:
It is only about a "minimal constraint", i.e. this PR can say that it at least requires JUPnP 2.1.0 (and this is the case, if I get you right). Any newer version of JUPnP will also be accepted then.

hmerk commented:
Ok, i will check this later this evening.

maggu2810 commented:
> It is only about a "minimal constraint", i.e. this PR can say that it at least requires JUPnP 2.1.0 (and this is the case, if I get you right). Any newer version of JUPnP will also be accepted then.

I would prefer using an "unbounded" version range, too. So we donate version x.y.z and all later versions.

hmerk commented:
@kaikreuzer @maggu2810 hopefully that's what you both wanted :-)

kaikreuzer commented:
Hopefully this is what you want as well :-D
Now you only need to squash once more!

hmerk commented:
For shure, just wanted to give you the chance to see this as a separate commit.
Squash will be done this evening...

hmerk commented:
@kaikreuzer 
Done !!!

kaikreuzer commented:
I was just about to merge, but having a closer look at the code for the first time, I am afraid that I still have some review comments...

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoDiscoveryParticipant.java:
```diff
@@ -0,0 +1,115 @@
+        if (device != null) {
+            if (device.getDetails().getManufacturerDetails().getManufacturer() != null) {
+                if (device.getDetails().getManufacturerDetails().getManufacturer().toUpperCase().contains("BELKIN")) {
+                    if (device.getDetails().getModelDetails().getModelName().toLowerCase().startsWith("socket")) {
```
I have seen UPnP devices where `getDetails().getModelDetails().getModelName()` leads to an NPE, because they lack the information - it is therefore better to do rigid null checks on such paths (or catch NPEs around the whole code block)

hmerk replied:
OK, will be fixed

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java:
```diff
@@ -0,0 +1,273 @@
+import org.eclipse.smarthome.io.transport.upnp.UpnpIOService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.CharacterData;
```
you need to also import org.w3c.dom and org.xml.sax in your manifest!

hmerk replied:
OK, will be added to manifest

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java:
```diff
@@ -0,0 +1,273 @@
+                                }
+
+                            } else {
+                                logger.info("Discovered an unsupported device :");
```
please only do this as debug

hmerk replied:
OK

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java:
```diff
@@ -0,0 +1,273 @@
+
+    @Override
+    public void onValueReceived(String variable, String value, String service) {
+        // TODO Auto-generated method stub
```
please remove the TODOs

hmerk replied:
OK

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoBridgeHandler.java:
```diff
@@ -0,0 +1,89 @@
+            for (Thing thing : getThing().getThings()) {
+                ThingHandler handler = thing.getHandler();
+                if (handler != null) {
+                    handler.initialize();
```
this looks BAAAAD - imho, a handler must never ever call initialize() on other handlers - this is something the framework takes care of.

hmerk replied:
I Understand, but I am quite sure that I found this in another binding where I copued this from.
The reason for doing this was that the thingHandlers got initialised before the bridgeHandler which throws errors and ther was now update later on.
I will see how to fix this.

kaikreuzer replied:
You probably need to make use of the bridgeHandlerInitialized and bridgeHandlerDisposed methods - and do less within initialize() of your thingHandler.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoMakerHandler.java:
```diff
@@ -0,0 +1,345 @@
+
+    @SuppressWarnings("unused")
+    private synchronized void onSubscription() {
+        // TODO Auto-generated method stub
```
remove TODOs

hmerk replied:
OK

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java:
```diff
@@ -56,151 +55,140 @@
+    protected final static int SUBSCRIPTION_DURATION = 600;
+
+    private UpnpIOService service;
+    private DiscoveryServiceRegistry discoveryServiceRegistry;
```
The DiscoveryServiceRegistry is not meant to be used by bindings.

If I see it right, you only use it to know if your device with a given UDN is available or not - this information should actually be provided by UpnpIOParticipant .onStatusChanged (which you have ignored in the code below).

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoLightHandler.java:
```diff
@@ -0,0 +1,360 @@
+                            break;
+                    }
+                } else if (command instanceof IncreaseDecreaseType) {
+                    // ToDo : not implemented yet
```
why not?

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoLightHandler.java:
```diff
@@ -0,0 +1,360 @@
+
+            if (getWemoBridgeHandler() != null) {
+                ThingStatusInfo statusInfo = getBridge().getStatusInfo();
+                updateStatus(statusInfo.getStatus(), statusInfo.getStatusDetail(), statusInfo.getDescription());
```
you should not copy the status of the bridge with all the details (which might not make sense for the device).
Note that the framework automatically sets your status to OFFLINE, if your bridge is OFFLINE. What is expected from you is to see if your device is ready to work and set it to ONLINE or if there is a problem (while the bridge being ONLINE), set it to OFFLINE with an according message (which only you can know here).

hmerk replied:
This is exactly the same like in the HueLightHandler.
So why is it ok there, but not here ??

kaikreuzer replied:
Who claims that it is ok there? Sounds to me like a bug to be fixed in the Hue binding then :-)

hmerk replied:
Well,I thought that the Hue binding was deeply reviewed and that this part of the code was ok.
If it is not, how should I know :-)

kaikreuzer replied:
I do not blame you. Only myself for not having done a proper review of all Hue binding changes then :-)

maggu2810 replied:
I have created https://github.com/eclipse/smarthome/issues/991 to keep track of the changes that should be done in the Hue binding.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoMakerHandler.java:
```diff
@@ -0,0 +1,345 @@
+    public final static Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_MAKER);
+
+    private UpnpIOService service;
+    private DiscoveryServiceRegistry discoveryServiceRegistry;
```
same comment as above - please do not use the DiscoveryServiceRegistry

kaikreuzer commented:
Sorry for coming back so late with comments, but I think there are a few things concerning the architecture that need to be changed before merging - hoping for your understanding!

hmerk commented:
@kaikreuzer 
I will try to fix the issues over the weekend, but there is at least one comment that I don't understand.
"you should not copy the status of the bridge with all the details (which might not make sense for the device)."
What I did was exactly the same like unbd the HueLightHandler.
So what's the difference ?????

kaikreuzer commented:
@hmerk ping :-)

hmerk commented:
@kaikreuzer 
;-) I am just updating the Readme and will then be ready to push my changes.

hmerk commented:
@kaikreuzer 
Done an ready for next review ;-)

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml:
```diff
@@ -16,7 +16,7 @@
                 <label>Unique Device Name</label>
                 <description>The UDN identifies the WeMo Device</description>
-		<required>true</required>
+        <required>true</required>
```
something screwed up with the formatting here

hmerk replied:
Fixed

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/OSGI-INF/WemoHandlerFactory.xml:
```diff
@@ -9,11 +9,10 @@
    <implementation class="org.eclipse.smarthome.binding.wemo.internal.WemoHandlerFactory"/>
-
+   <reference bind="setUpnpIOService" cardinality="1..1" interface="org.eclipse.smarthome.io.transport.upnp.UpnpIOService" name="UpnpIOService" policy="static" unbind="unsetUpnpIOService"/>
+   <reference bind="setDiscoveryServiceRegistry" cardinality="1..1" interface="org.eclipse.smarthome.config.discovery.DiscoveryServiceRegistry" name="DiscoveryServiceRegistry" policy="static" unbind="unsetDiscoveryServiceRegistry"/>
```
this line is obsolete now, isn't it?

hmerk replied:
Indeed, was a left ofer, fixed

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoMakerHandler.java:
```diff
@@ -0,0 +1,324 @@
+ * @author Hans-Jörg Merk - Initial contribution
+ */
+
+public class WemoMakerHandler extends BaseThingHandler implements UpnpIOParticipant, DiscoveryListener {
```
Do you really really need to be a DiscoveryListener?
As per my previous comment, I think `UpnpIOParticipant.onStatusChanged()` should give you all that you need.

hmerk replied:
Hmm, not sure about it.
I then cannot use thingDiscovered() and thingRemoved() methods.
If it's not a problem, I would leave this for now.

kaikreuzer replied:
Why are thingDiscovered() and thingRemoved() any better than onStatusChanged()? This method is supposed to do exactly what you wish for or am I missing something?

kaikreuzer commented:
Thanks @hmerk. Review done, back to author ;-)

hmerk commented:
Thanks @kaikreuzer, ang back again :-)

kaikreuzer commented:
@hmerk Not sure if you saw my https://github.com/eclipse/smarthome/pull/721/files#r53696551? This is the last point to clarify - sorry if I don't get the point there yet.

hmerk commented:
@kaikreuzer, yes, I saw it this morning and will fix this asap, but won't be able to look into it before the weekend.
Lot of travelling again this week...

hmerk commented:
@kaikreuzer please have a look at latest commit.
I removed the DiscoveryListener and added statuslistener and onStatusChanged().

hmerk commented:
@kaikreuzer 
If my latest commit is fine for you, I would prefer to merge this as a separate commit.
As You know, the WeMo devices react a bit strange on using UPnP actions, thats why I am doing my own HTTP calls. If we encounter problems while using UPnP for queriing the device reachability, we could simply revert this commit and fall back to the thingDiscovered() methods.
I hope that this will not be needed, as the latest version is running fine for about 20 minutes in my IDE now.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoLinkDiscoveryService.java:
```diff
@@ -0,0 +1,269 @@
+ * @author Hans-Jörg Merk - Initial contribution
+ *
+ */
+public class WemoLinkDiscoveryService extends AbstractDiscoveryService implements UpnpIOParticipant {
```
Why do you implement UpnpIOParticipant if you do not implement onValueReceived nor onStatusChanged? Can't this be removed then?

hmerk replied:
Seems to be a leftover from my different tries, I will check this tonight and solve it.

hmerk replied:
Should not answer to quick. UpnpIOParticipant is needed cause I have to use getDescriptorURL() for retreiving the IP-Adress of the bridge.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoHandler.java:
```diff
@@ -56,151 +49,113 @@
-            logger.debug("Initializing WemoHandler for UDN '{}'", configuration.get(UDN));
+        if (configuration.get("udn") != null) {
+            logger.debug("Initializing WemoHandler for UDN '{}'", configuration.get("udn"));
+            if (service.isRegistered(this)) {
```
So if it is not registered at the time that the handler is initialized, you will never add a status listener then?

hmerk replied:
You are right, i will move it so the listener will be registered always.

kaikreuzer commented:
Hi @hmerk - only two questions left :-)

hmerk commented:
Hi Kai,
hopefully you will not find more issues after I fix these two ;-)
Will try to commit the corrections tonight.

hmerk commented:
@kaikreuzer please see my comment on using UpnpIOParticipant can't be removed.
Changed call of addStatusListener().
I hope everything is ok now, otherwise, let me know.

kaikreuzer commented:
Thanks, @hmerk.
I've now tried to do a functional test, but had problems there.
But first two remarks on the code:
- In my IDE, I have an error on the jupnp 2.1.0 import. Since the jupnp bundle does not export the package with a version, it seems that this cannot be resolved here. Is it working for you?
- Rebasing on the latest master gives me deprecation warnings in WemoHandler (due to some changes on the ChannelUID class) - maybe you could adapt this?

Regarding the functional testing: My WeMo switch fails to poll its state. In the log I get
```
2016-02-25 11:42:59.845 [DEBUG] [e.s.i.t.upnp.UpnpIOServiceImpl:400  ] - Polling participant 'Socket-1_0-221318K110066D' through Action 'GetBinaryState' of Service 'basicevent1' 
Feb 25, 2016 11:42:59 AM org.jupnp.transport.spi.AbstractStreamClient sendRequest
WARNING: HTTP request failed: (OutgoingActionRequestMessage) POST http://192.168.188.20:49153/upnp/control/basicevent1
org.apache.http.NoHttpResponseException: 192.168.188.20:49153 failed to respond
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:143)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:57)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:260)
	at org.apache.http.impl.AbstractHttpClientConnection.receiveResponseHeader(AbstractHttpClientConnection.java:283)
	at org.apache.http.impl.conn.DefaultClientConnection.receiveResponseHeader(DefaultClientConnection.java:251)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.receiveResponseHeader(ManagedClientConnectionImpl.java:197)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:271)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:123)
	at org.apache.http.impl.client.DefaultRequestDirector.tryExecute(DefaultRequestDirector.java:685)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:487)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:863)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:72)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:214)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:160)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:136)
	at org.jupnp.transport.impl.apache.StreamClientImpl$4.call(StreamClientImpl.java:171)
	at org.jupnp.transport.impl.apache.StreamClientImpl$4.call(StreamClientImpl.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

2016-02-25 11:42:59.852 [DEBUG] [e.s.i.t.upnp.UpnpIOServiceImpl:411  ] - Signalling that 'Socket-1_0-221318K110066D' is not responding
2016-02-25 11:42:59.853 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'wemo:socket:Socket-1_0-221318K110066D' updated: OFFLINE
2016-02-25 11:42:59.853 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'wemo:socket:Socket-1_0-221318K110066D' changed from ONLINE to OFFLINE
```
Afterwards, it does not retry polling anymore, so it stays OFFLINE forever.
Manually trying to access `http://192.168.188.20:49153/upnp/control/basicevent1` results in a 404. Any idea what might be wrong here?

hmerk commented:
@kaikreuzer
JUPnP -> When testing, I also had a problem with the Version constraint but thought its my IDE, easiest way would be to remove it from manifest.
 ChannelUID class -> Did not see this in my IDE, will try to reproduce and see what's wrong.
Functional Test -> That's exactly what I commented regarding use of ThingDiscovered() and ThingRemoved().
The newly introduced addStatuslistener() and onChanged() make use of the UPnP actions, which are not really relaiable in case of using JUPnP, that's wha I don't use these in handleCommand(9, but make my own HTTP calls.
Therefore I left ist as a separate commit, so we can decide wether to use this changes or not.
If you like, we can discuss this by phone, I am available today :-)

kaikreuzer commented:
Ok, I think I see your point. I was hoping that UPnPIOService works reliably, but as its "presence detection" is using UPnP actions and they somehow do not work well with WeMo, this simply does not work.
So I am fine with your workaround then (and I just tested it successfully). Please don't forget to still include the other changes that you did in the second commit as well.
Please squash your stuff afterwards and we should be good to merge - thanks for your patience!

hmerk commented:
It's not UPnPIOService causing the problem, it's WeMo :-(
There should be no changes in the second commit that need to be transfered to the first on, but I will check again.

kaikreuzer commented:
> There should be no changes in the second commit that need to be transfered to the first on

I guess you are right, sorry.

hmerk commented:
Ok, just checked, nothing to be transfered.

hmerk commented:
@kaikreuzer 
What about the other two questions you asked (ChannelUID class and JUPnP Version constraint) ?
Is this still something I should look at ?

kaikreuzer commented:
Yes, please, that should be quick to do!

hmerk commented:
@kaikreuzer ok, think I am ready now:
reverted commit as discussed
removed JUPnP version constraint
Updated use of ChannelUID class
Squashed commits

Anything else ???? :-)))))))

kaikreuzer commented:
> Anything else ???? :-)))))))

No, why do you ask? Did you expect any further comments from me? :stuck_out_tongue:

hmerk commented:
:-)
